### ISSUE TITLE
NullReferenceException on Timeout Expired error during execution of DeleteMetadata command

### ISSUE THREAD

**Issue by kibao** [Mimetis/Dotmim.Sync#873]
### Description

No exception is thrown on `Timeout Expired` error during execution of DeleteMetadata command and then `syncRowCountParam` is not null, but `syncCountParam.Value` is null which leads to `NullReferenceException` during casting to `int`.

This part passes exception through:
https://github.com/Mimetis/Dotmim.Sync/blob/1fd224abbe8d5f3fc02d0bbc08f5a7ee5a9f58d4/Projects/Dotmim.Sync.Core/Orchestrators/Metadatas/BaseOrchestrator.Metadatas.cs#L77-L86
<details>
<summary>Exception</summary>

```
Microsoft.Data.SqlClient.SqlException (0x80131904): Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.
 ---> System.ComponentModel.Win32Exception (258): Unknown error: 258
   at Microsoft.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at Microsoft.Data.SqlClient.TdsParserStateObject.ThrowExceptionAndWarning(Boolean callerHasConnectionLock, Boolean asyncClose)
   at Microsoft.Data.SqlClient.TdsParserStateObject.CheckThrowSNIException()
   at Microsoft.Data.SqlClient.SqlCommand.CheckThrowSNIException()
   at Microsoft.Data.SqlClient.SqlCommand.InternalEndExecuteNonQuery(IAsyncResult asyncResult, Boolean isInternal, String endMethod)
   at Microsoft.Data.SqlClient.SqlCommand.EndExecuteNonQueryInternal(IAsyncResult asyncResult)
   at Microsoft.Data.SqlClient.SqlCommand.EndExecuteNonQueryAsync(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location ---
   at Dotmim.Sync.BaseOrchestrator.InternalDeleteMetadatasAsync(IEnumerable`1 scopeInfos, SyncContext context, Int64 timestampLimit, DbConnection connection, DbTransaction transaction, CancellationToken cancellationToken, IProgress`1 progress)
ClientConnectionId:25d3a285-10cf-42e3-afa8-efb45f36aa27
Error Number:-2,State:0,Class:11
```
</details>

Then casting throws `NullReferenceException`:
https://github.com/Mimetis/Dotmim.Sync/blob/1fd224abbe8d5f3fc02d0bbc08f5a7ee5a9f58d4/Projects/Dotmim.Sync.Core/Orchestrators/Metadatas/BaseOrchestrator.Metadatas.cs#L88-L92
<details>
<summary>Exception</summary>

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Dotmim.Sync.BaseOrchestrator.InternalDeleteMetadatasAsync(IEnumerable`1 scopeInfos, SyncContext context, Int64 timestampLimit, DbConnection connection, DbTransaction transaction, CancellationToken cancellationToken, IProgress`1 progress)
```
</details>

#### Simple - For existing setup

Requirements:
- Local database: SQL Server

1. Add `WAITFOR DELAY '00:00:10';` to corresponding `_deletemetadata` procedure on `local` database.
2. Set `SyncOptions.DbCommandTimeout` to 1 sec.
3. Change some rows in table on `remote` database,
4. Run `SynchronizeAsync`

#### Advanced with large data set.

In my case, it happened during second `SynchronizeAsync` (where first one was initial for existing tables).

Remote database: SQL Server
Local database: SQL Server
Set `SyncOptions.DbCommandTimeout` to 1 sec.

Steps:
1. Create table on `remote` database with 2'000'000 rows.
2. Run initial `SynchronizeAsync`.
    As result, corresponding tracking table on `local` database has 2'000'000 rows.
4. Change some rows in table on `remote` database,
5. Run second `SynchronizeAsync`
    It executes DeleteMetadata from `BaseOrchestrator.InternalDeleteMetadatasAsync` which times out.

### Possible solution

1. Add check whether `syncCountParam.Value` is not null before casting
2. Log exception for debugging. 
3. Maybe throw SyncException when Exception is not related to missing stored procedure.

**Comment by Mimetis** [comment]
Thanks for (once again) a great detailed issue (OMG, why everything is not so well documented :))
Indeed, the exception here should be raised. I guess the stored procedure exception is some kind of historical thing.
I think a classic throw error should be fine here.
Just removing the try catch should do the trick

**Comment by Mimetis** [comment]
Once again, thanks for the detailed feedback
It was **INCREDIBLY EASY TO REPRODUCE** thanks **TO YOUR EFFORT TO MAKE A REPRO SAMPLE**

_God... I wish all issues would be so well documented..._

Anyway, here are my thoughts:

Now, as you said we have a weird error that do not help to figure out what is happening:

``` cs
System.NullReferenceException: Object reference not set to an instance of an object
```

Removing the try catch (I think it's historical and should have been removed before) will raise the error to the top stack:
``` cs
//try
//{
    rowsCleanedCount = await command.ExecuteNonQueryAsync().ConfigureAwait(false);
//}
//catch (Exception)
//{
//}
```

Exception:
``` cs
Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.
```

It's better, but it's still hard to figure out the error is raised by the `deletemetadata` process

Usually for a lot of know issues, I used to create my own typed exception, for example

``` cs
/// <summary>
/// Occurs when trying to use a closed connection
/// </summary>
public class ConnectionClosedException : Exception
{
    const string message = "The connection to database {0} is closed.";

    public ConnectionClosedException(DbConnection connection) : base(string.Format(message, connection.Database)) { }
}
```

But once again, the error raised here is coming from the database and could be raised from any method.

I working on a new way to raise error.
The idea is to have the full stack trace with every methods called to the very last one who raised the error:

``` sh
Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding. - [(localdb)\mssqllocaldb]\[Client]
├─ Error raised from InternalApplyChangesAsync. See Inner Exceptions for details
├── Error raised from InternalDeleteMetadatasAsync. See Inner Exceptions for details
├─── Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.
└──── The wait operation timed out.
```
Of course this level of debug information would be available if you are setting the `options.UseVerboseErrors=true;`

It's involving a bit of work to be sure that all methods are wrapped with this kind of "_new error handling_"

What do you think ?

**Comment by Mimetis** [comment]
Available in the new version `0.9.8` (available as a beta version atm)
Thanks again @kibao for your great feedback and detailed explanation
